bindings/go/descriptor/descriptor.go

```diff
@@ -0,0 +1,177 @@
+	}
+	return base
+}
+
+type Component struct {
+	ComponentIdentity  `json:",inline"`
+	Labels             []*Label                `json:"labels,omitempty"`
+	RepositoryContexts []*runtime.Unstructured `json:"repositoryContexts,omitempty"`
+	Provider           string                  `json:"provider"`
+	Resources          []*Resource             `json:"resources,omitempty"`
+	Sources            []*Source               `json:"sources,omitempty"`
+	References         []*Reference            `json:"componentReferences,omitempty"`
+}
+
+type ComponentIdentity struct {
+	Name    string `json:"name"`
+	Version string `json:"version"`
+}
+
+func (c *ComponentIdentity) String() string {
+	return fmt.Sprintf("%s:%s", c.Name, c.Version)
+}
+
+func ParseComponentIdentity(s string) (ComponentIdentity, error) {
+	ci := ComponentIdentity{}
+	split := strings.Split(s, ":")
+	if len(split) != 2 {
+		return ci, fmt.Errorf("invalid component identity %q", s)
+	}
+	ci.Name = split[0]
+	ci.Version = split[1]
+	if ci.Name == "" || ci.Version == "" {
+		return ci, fmt.Errorf("invalid component identity %q", s)
+	}
+	return ci, nil
+}
+
+func (c *ComponentIdentity) AsIdentity() runtime.Identity {
+	return runtime.Identity{
+		"name":    c.Name,
+		"version": c.Version,
+	}
+}
+
+type Resource struct {
+	ObjectMeta    `json:",inline"`
+	ExtraIdentity map[string]string `json:"extraIdentity,omitempty"`
+	SourceRefs    []*SourceRef      `json:"sourceRefs,omitempty"`
+	Type          string            `json:"type"`
+	Relation      string            `json:"relation"`
+	Access        Access            `json:"access"`
+	Digest        *Digest           `json:"digest,omitempty"`
+	Size          int64             `json:"size,omitempty"`
+}
+
+func (r *Resource) GetIdentity() map[string]string {
+	m := maps.Clone(r.ExtraIdentity)
+	if m == nil {
+		m = make(map[string]string)
+	}
+	m["name"] = r.Name
+	return m
+}
+
+type Source struct {
+	ObjectMeta    `json:",inline"`
+	ExtraIdentity map[string]string `json:"extraIdentity,omitempty"`
+	Type          string            `json:"type"`
+	Access        Access            `json:"access"`
+}
+
+func (r *Source) GetIdentity() map[string]string {
+	m := maps.Clone(r.ExtraIdentity)
+	if m == nil {
+		m = make(map[string]string)
+	}
+	m["name"] = r.Name
+	return m
+}
+
+type Access struct {
+	runtime.Typed `json:",inline"`
+}
+
+func (a *Access) UnmarshalJSON(data []byte) error {
+	a.Typed = &runtime.Raw{}
+	return json.Unmarshal(data, a.Typed)
+}
+
+func (a *Access) MarshalJSON() ([]byte, error) {
+	return json.Marshal(a.Typed)
+}
+
+type Reference struct {
+	Name          string            `json:"name"`
+	ExtraIdentity map[string]string `json:"extraIdentity,omitempty"`
+	Component     string            `json:"componentName"`
+	Version       string            `json:"version"`
+	Digest        Digest            `json:"digest,omitempty"`
+	Labels        []*Label          `json:"labels,omitempty"`
```

Comment:
I think slices of pointers are usually a bad idea

Reply:
Everything in here is slices of pointers. Do you want to change everything?